Dedicated etcd version check

ch1bo opened the pull request:
This checks the version compiled into `hydra-node` against a version stored on the `etcd` key-value store and emits a "network-level" version mismatch `ConnectivityEvent` whenever we (re-)connect to the majority cluster.

This is different to before where we would encode a version into each peers message (quite crudely) and also check it on each message received.

---

* [x] CHANGELOG updated
* [x] Documentation update not needed
* [x] Haddocks updated
* [ ] New TODOs explained herafter
  - Encountered that the `prop_specIsComplete` used in `ClientInputSpec` and `ServerOutputSpec` is not working!

ch1bo committed:
Draft NetworkVersionMismatch types

Shifting to a network-level incompatibility message makes it easier to
determine (especially for etcd based network).

In the end, an operator only needs to know whether they need to update
the hydra-node or their counter parties.

ch1bo committed:
Update schemas and fix schema tests

Apparently the prop_specIsComplete is not sufficient to detect
incompatibilities (and seems to be broken too?).

ch1bo committed:
Compare and swap of version via a txn request

ch1bo committed:
Also report NetworkMismatch in case version is not parsable

The version check test is still flaky.

noonio commented on hydra-node/src/Hydra/Network/Etcd.hs:
```diff
@@ -133,16 +135,18 @@ withEtcdNetwork tracer protocolVersion config callback action = do
-                    }
-                atomically (writeTVar doneVar True)
+        withConnection (connParams doneVar) grpcServer $ \conn ->
+          -- REVIEW: checkVersion blocks if used on main thread - why?
```
Seems to make sense; it directly connects and waits for a response?

ch1bo replied:
Does not explain anything for me. It blocks forever on the main thread.. why would it? One Hypothesis is that the connection is not established correctly (done lazily using first rpc call) if we only try `checkVersion` and none of the other requests.

noonio replied:
Oh, I didn't see that you meant _blocks forever_. That is a problem.

noonio replied:
Investigated this for a bit and decided to not worry for now.

noonio commented on hydra-node/src/Hydra/Network/Etcd.hs:
```diff
@@ -227,46 +231,62 @@ withEtcdNetwork tracer protocolVersion config callback action = do
+
+  -- Check version if version was already present
+  if res ^. #succeeded
+    then forM_ (res ^.. #responses . traverse . #responseRange . #kvs . traverse) $ \kv ->
```
I worry the `forM_` here will silently not emit warnings if there's no response? Is that what we want?

ch1bo replied:
It would mean that the `etcd` grpc server did not follow its promise of the etcd API (matching `txn` requests with responses). As this is not strictly mission critical (the version check is more FYI), I would not complicate this further.

noonio commented on hydra-node/test/Hydra/NetworkSpec.hs:
```diff
@@ -112,47 +111,48 @@ spec = do
+                    aliceSeesMismatch = waitEq waitAlice 5 mismatch
+                -- Hence, either one or the other will see the mismatch
+                -- FIXME: this is still flaky, but less so than before
+                race_ bobSeesMismatch aliceSeesMismatch
```
Note that, as implemented, you could test a version mismatch between one peer.

Maybe that's easier?

ch1bo replied:
It was just wrong, fixed now.

ch1bo committed:
Fix assertion in NetworkSpec

ch1bo committed:
Serialize Nothing fields as null for ServerOutput

The NetworkVersionMismatch may or may not know what 'theirVersion' is
and this is indicated by 'null' or an integer.

ch1bo committed:
Fix logs.yaml schema
